### PR TITLE
Add migrationOptions.resolveEmbeddedConstraints

### LINKED ISSUE #6073: Orphaned embedded objects are not deleted when migrating from Object to Embedded Object
### How frequently does the bug occur?

Always

### Description

This issue has previously been noted and resolved in realm-java and realm-kotlin:

https://github.com/realm/realm-java/issues/7769
https://github.com/realm/realm-kotlin/issues/1464

Orphaned documents should be handled automatically according to this change in realm-core:
https://github.com/realm/realm-core/pull/5737

The issue occurs when migrating a schema for a local realm. Version 2 of the schema has three object models defined. Version 3 of the schema embeds each of these three object models into one another in order to take advantage of cascading deletes. When the existing realm is attempted to be migrated from version 2 to 3, the following error is received:

"Failed to open local Realm for {REALM_INSTANCE} - Error: ","args":["Cannot convert '{OBJECT}' to embedded: at least one object has no incoming links and would be deleted."]

### Stacktrace & log output

```shell
"Failed to open local Realm for {REALM_INSTANCE} - Error: ","args":["Cannot convert '{OBJECT}' to embedded: at least one object has no incoming links and would be deleted."]
```

### Can you reproduce the bug?

Always

### Reproduction Steps

_No response_

### Version

11.3.1

### What services are you using?

Local Database only

### Are you using encryption?

-- select --

### Platform OS and version(s)

Electron 22.3.14

### Build environment

Which debugger for React Native: ..

### Cocoapods version

_No response_

### PR DESCRIPTION
**Pull request by kneth**
## What, How & Why?

This closes #6073 

## ☑️ ToDos

* [x] 📝 Changelog entry
* ~[ ] 📝 `Compatibility` label is updated or copied from previous entry~
* ~[ ] 📝 Update `COMPATIBILITY.md`~
* [x] 🚦 Tests
* ~[ ] 📦 Updated internal package version in consuming `package.json`s (if updating internal packages)~
* ~[ ] 📱 Check the React Native/other sample apps work if necessary~
* ~[ ] 💥 `Breaking` label has been applied or is not necessary~

### CODE REVIEW

File: integration-tests/tests/src/tests/migrations.ts [hunk 1]
```diff
@@ -521,4 +521,75 @@ describe("Migrations", () => {
...
+      });
+
+      realm.close();
+    });
+
+    it("migration should delete orphan objects", () => {
+      const realm = new Realm({
+        schema: [ParentSchema, EmbeddedChildSchema],
+        schemaVersion: 2,
+        automaticallyHandleBacklinksInMigration: true,
```
Comment (jcarpe) [inline integration-tests/tests/src/tests/migrations.ts:577]: 👏

File: packages/realm/src/Configuration.ts [hunk 2]
```diff
@@ -47,6 +50,15 @@ export type BaseConfiguration = {
   disableFormatUpgrade?: boolean;
   encryptionKey?: ArrayBuffer | ArrayBufferView | Int8Array;
   onMigration?: MigrationCallback;
+  /**
+   * A flag to indicate whether Realm should resolve
+   * embedded object constraints after migration. If this is `true` then all embedded objects
+   * without a parent will be deleted and every embedded object with multiple references to it
```
Comment (jcarpe) [inline packages/realm/src/Configuration.ts:56]: Nit - is it "multiple"? Or would this apply even for a single reference (I think as the test implies)? May suggest saying: "...every embedded object with one or more references..."

File: packages/realm/src/Configuration.ts [hunk 3]
```diff
@@ -47,6 +50,15 @@ export type BaseConfiguration = {
...
+   * will be duplicated so that every referencing object will hold its own copy of the
+   * embedded object.
+   * @default false
+   */
```
Comment (takameyer) [inline packages/realm/src/Configuration.ts:60]: I'm noticing that this seems to be the only configuration option that is documented.  I believe we need a follow-up ticket to document the realm configuration in code.

**Review by takameyer — APPROVED** [review]
LGTM :shipit:

File: packages/realm/src/Configuration.ts [hunk 4]
```diff
@@ -47,6 +50,15 @@ export type BaseConfiguration = {
...
+  automaticallyHandleBacklinksInMigration?: boolean;
```
Comment (elle-j) [inline packages/realm/src/Configuration.ts:61]: Wondering if it'd be better to name this more in line with how you're describing it in the docs. Perhaps:
* `autoResolveEmbeddedConstraintsInMigration`
* `autoHandleEmbeddedConstraintsInMigration`
* (ps. Core uses the plural of Migration.)

Thoughts?

**Review by elle-j — APPROVED** [review]
Nice! Only got a naming related comment.

File: packages/realm/src/Configuration.ts (hunk 3 above)
Comment (kneth) [reply packages/realm/src/Configuration.ts:60]: Yep, I noticed too. I have created https://github.com/realm/realm-js/issues/6081

File: packages/realm/src/Configuration.ts (hunk 2 above)
Comment (kneth) [reply packages/realm/src/Configuration.ts:56]: I plead guilty: I copied from Kotlin :smile:

Your formulation sounds more accurate, and I will update before merging.

File: packages/realm/src/Configuration.ts (hunk 4 above)
Comment (elle-j) [reply packages/realm/src/Configuration.ts:61]: Sounds good to me 👍

File: packages/realm/src/Configuration.ts [hunk 5]
```diff
@@ -109,6 +130,7 @@ export function validateConfiguration(config: unknown): asserts config is Config
   }
   if (sync !== undefined) {
     assert(!onMigration, "The realm configuration options 'onMigration' and 'sync' cannot both be defined.");
+    assert(!migrationOptions, "The realm configuration options 'migrationOptions' and 'sync' cannot both be defined.")
```
Comment (takameyer) [inline packages/realm/src/Configuration.ts:133]:
```suggestion
    assert(!migrationOptions, "The realm configuration options 'migrationOptions' and 'sync' cannot both be defined.");
```
;)

**Review by takameyer — APPROVED** [review]
:shipit:

File: packages/realm/src/Realm.ts [hunk 6]
```diff
@@ -547,6 +547,9 @@ export class Realm {
         encryptionKey: Realm.determineEncryptionKey(config.encryptionKey),
         syncConfig: config.sync ? toBindingSyncConfig(config.sync) : undefined,
         forceSyncHistory: config.openSyncedRealmLocally,
+        automaticallyHandleBacklinksInMigrations: config.migrationOptions
+          ? config.migrationOptions.resolveEmbeddedConstraints === true
+          : false,
```
Comment (elle-j) [inline packages/realm/src/Realm.ts:552]: One alternative. (Feel free to choose whichever. If going with the one below, double check linting is happy 🙂)
```suggestion
        automaticallyHandleBacklinksInMigrations: config.migrationOptions?.resolveEmbeddedConstraints ?? false,
```

**Review by elle-j — APPROVED** [review]

**Comment by elle-j** [comment]
Addition: Also export `MigrationOptions` in [index.ts](https://github.com/realm/realm-js/blob/4715caa2283d8903131513bb635026995168e410/packages/realm/src/index.ts#L108) and on the [Realm namespace](https://github.com/realm/realm-js/blob/4715caa2283d8903131513bb635026995168e410/packages/realm/src/Realm.ts#L1462).

**Pull request merged**